atc/wrappa/concurrent_request_limits_wrappa.go

```diff
@@ -0,0 +1,48 @@
+package wrappa
+
+import (
+	"net/http"
+
+	"code.cloudfoundry.org/lager"
+
+	"github.com/tedsuo/rata"
+)
+
+//go:generate counterfeiter code.cloudfoundry.org/lager.Logger
+
+type ConcurrencyLimitsWrappa struct {
+	logger                  lager.Logger
+	concurrentRequestPolicy ConcurrentRequestPolicy
+}
+
+func NewConcurrencyLimitsWrappa(
+	logger lager.Logger,
+	concurrentRequestPolicy ConcurrentRequestPolicy,
+) Wrappa {
+	return ConcurrencyLimitsWrappa{
+		logger:                  logger,
+		concurrentRequestPolicy: concurrentRequestPolicy,
+	}
+}
+
+func (wrappa ConcurrencyLimitsWrappa) Wrap(handlers rata.Handlers) rata.Handlers {
+	wrapped := rata.Handlers{}
+
+	for action, handler := range handlers {
+		if wrappa.concurrentRequestPolicy.IsLimited(action) {
+			pool := wrappa.concurrentRequestPolicy.HandlerPool(action)
+			wrapped[action] = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				if !pool.TryAcquire() {
+					wrappa.logger.Info("concurrent-request-limit-reached")
+					w.WriteHeader(http.StatusTooManyRequests)
```

Comment:
I know this was discussed somewhere already, but is a 429 the right status code for this? In theory a user could submit a single request and the api responds telling them to rate limit themselves, which seems confusing.

I this this as more of a circuit breaker pattern, where the service would typically respond with a 503 Service Unavailable.

Reply:
hmmm [the rfc defining the 429 code](https://tools.ietf.org/html/rfc6585#section-4) does pretty clearly indicate that "_the user_ has sent too many requests in a given amount of time" and this condition is not scoped or dependent on any particular user.

503, on the other hand, [says](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503) "Common causes are a server ... that is overloaded" and also allows for a `Retry-After` header, should we eventually choose to include one. It does sound like a better fit.

At first I was concerned about the web UI potentially giving special treatment to a 500-series status code, which anecdotally means "terrible bad stuff you can't possibly understand", but 503 seems like a scenario where retries are normal.